### PR TITLE
Change ABIEncoderV1 to not pad empty strings

### LINKED ISSUE #10170: Extra bytes returned on revert("")
## Description

Calling `revert(string)` should return abi-encoded string as if it were a call to `Error(string)` function ([docs](https://solidity.readthedocs.io/en/latest/control-structures.html#revert)).

Therefore when calling revert with empty string `revert("")` I expect to receive correct abi-encoded `Error("")`: 
```
0x08c379a0                                                          <- Error(string) selector
0000000000000000000000000000000000000000000000000000000000000020    <- string offset = 32
0000000000000000000000000000000000000000000000000000000000000000    <- string length = 0
```

In fact Solidity appends extra 32 bytes of zeroes in that case so returned bytes are
```
0x08c379a0                                                          <- Error(string) selector
0000000000000000000000000000000000000000000000000000000000000020    <- string offset = 32
0000000000000000000000000000000000000000000000000000000000000000    <- string length = 0
0000000000000000000000000000000000000000000000000000000000000000    <- unexpected extra 32 bytes
```

## Environment

- Compiler version: 0.6.12, 0.7.4
- Target EVM version: compiler default (Istanbul)
- Framework/IDE: Remix
- EVM: mainnet, ropsten, ganache-cli, ethereumjs-vm

## Steps to Reproduce

```solidity
contract RevertReasonParserTest {
    function emptyStringRevert() external pure  {
        revert("");
    }

    function test() external view returns(bytes memory, uint256) {
        try this.emptyStringRevert() {
        } catch (bytes memory reason) {
            return (reason, reason.length);
        }
    }
}
```

[Ropsten example](https://ropsten.etherscan.io/address/0xfdd9aa7ae91e966ae3d8aa708d5fe42f82ab746c)

[Mainnet example](https://etherscan.io/address/0x3a428d3ca52251fa40e4841e4556e13481893b83)

### PR DESCRIPTION
**Pull request by axic**
Fixes #10170.

Only the last commit here is what is relevant, the other commit may show up due to rebasing the dependant branch.

### CODE REVIEW

**Comment by axic** [comment]
Created this a quick test to see how to fix it, and it is surprisingly easy, but worrying that only a single test existed for this behaviour. Not sure if we should try to fix this, but opinions welcome.

File: libsolidity/codegen/CompilerUtils.cpp [hunk 1]
```diff
@@ -531,10 +531,14 @@ void CompilerUtils::encodeToMemory(
...
-				m_context << u256(strType.value().size());
+				auto const size = strType.value().size();
+				m_context << u256(size);
 				storeInMemoryDynamic(*TypeProvider::uint256(), true);
 				// stack: ... <end_of_mem'>
-				storeInMemoryDynamic(strType, _padToWordBoundaries);
+
+				// Do not output empty padding for zero-length strings.
+				if (size != 0)
+					storeInMemoryDynamic(strType, _padToWordBoundaries);
```
Comment (chriseth) [inline libsolidity/codegen/CompilerUtils.cpp:542]: Why not modify `storeInMemoryDynamic` instead? How does `string memory x = ""; revert(x);` behave?

Comment (axic) [reply libsolidity/codegen/CompilerUtils.cpp:542]: > How does string memory x = ""; revert(x); behave?

Apparently no padding in that case. See #10193.

Comment (leonardoalt) [reply libsolidity/codegen/CompilerUtils.cpp:542]: So the only affected type was `StringLiteral`?
If so, I think this is ok.

Comment (axic) [reply libsolidity/codegen/CompilerUtils.cpp:542]: If I move this check to `storeInMemoryDynamic` then this fails:
```
semanticTests/various/literal_empty_string.sol: FAIL
  Contract:
    contract C {
        bytes32 public x;
        uint256 public a;
    
        function f(bytes32 _x, uint256 _a) public {
            x = _x;
            a = _a;
        }
    
        function g() public {
            this.f("", 2);
        }
    }
    
    // ====
    // compileViaYul: also

  Expected result:
  // x() -> 0
  // a() -> 0
  // g() ->
  // x() -> 0
  // a() -> 2

  Obtained result:
  // x() -> 0
  // a() -> 0
  // g() -> FAILURE
  Warning: The call to "g()" returned 

  // x() -> 0
  // a() -> 0
  Warning: The call to "a()" returned 
  [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
  
  Attention: Updates on the test will apply the detected format displayed.
  
  Note that the test also has to pass via Yul.

(e)dit/(u)pdate expectations/(s)kip/(q)uit? 
```

Comment (axic) [reply libsolidity/codegen/CompilerUtils.cpp:542]: @chriseth this has been blocked on the above, I still didn't had the patience to debug this. Checked all` storeInMemoryDynamic` calls roughly, and none of them seemed suspect.

At least we should merge the comment improvements. We could merge this fix/improvement, and we could also try to find the reason -- however I'm not sure we need/have to do this.

**Comment by chriseth** [comment]
I'm fine with merging this in its current state.

File: libsolidity/codegen/CompilerUtils.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/codegen/CompilerUtils.cpp:542]: Moved this into an issue.

**Comment by axic** [comment]
@chriseth squash and rebased. If you are confident feel free to merge.

**Review by chriseth — APPROVED** [review]

**Pull request merged**